### PR TITLE
CORE-17355 - Reduce number of p2p e2e sessions maintained per pair of vnodes

### PR DESCRIPTION
**Pull request by nikinagy**
In order to increase our performance, this PR reduces number of p2p e2e sessions maintained per pair of vnodes:
- for member <-> MGM communication reduce to 1
- for member <-> member communication reduce to 1

### CODE REVIEW

File: components/link-manager/src/integrationTest/kotlin/net/corda/p2p/linkmanager/integration/LinkManagerIntegrationTest.kt [hunk 1]
```diff
@@ -112,7 +113,8 @@ class LinkManagerIntegrationTest {
             .withValue(MAX_REPLAYING_MESSAGES_PER_PEER, ConfigValueFactory.fromAnyRef(100))
             .withValue(HEARTBEAT_MESSAGE_PERIOD_KEY, ConfigValueFactory.fromAnyRef(2000))
             .withValue(SESSION_TIMEOUT_KEY, ConfigValueFactory.fromAnyRef(10000))
-            .withValue(SESSIONS_PER_PEER_KEY, ConfigValueFactory.fromAnyRef(4))
+            .withValue(SESSIONS_PER_PEER_FOR_MEMBER_KEY, ConfigValueFactory.fromAnyRef(4))
```
Comment (yift-r3) [inline components/link-manager/src/integrationTest/kotlin/net/corda/p2p/linkmanager/integration/LinkManagerIntegrationTest.kt]: 2?

File: components/link-manager/src/main/kotlin/net/corda/p2p/linkmanager/sessions/SessionManagerImpl.kt [hunk 2]
```diff
@@ -196,7 +197,8 @@ internal class SessionManagerImpl(
     @VisibleForTesting
     internal data class SessionManagerConfig(
         val maxMessageSize: Int,
-        val sessionsPerCounterparties: Int,
+        val sessionsPerCounterpartiesForMembers: Int,
```
Comment (yift-r3) [inline components/link-manager/src/main/kotlin/net/corda/p2p/linkmanager/sessions/SessionManagerImpl.kt]: Should we put this pair in its own data class?

File: components/link-manager/src/main/kotlin/net/corda/p2p/linkmanager/sessions/SessionManagerImpl.kt [hunk 3]
```diff
@@ -250,12 +253,35 @@ internal class SessionManagerImpl(
...
+            us.toCorda(),
+            peer.toCorda(),
+            status,
+            counterpartyInfo.serial,
+            isCommunicationBetweenMgmAndMember(ourInfo, counterpartyInfo)
+        )
+    }
+
+    private fun isCommunicationBetweenMgmAndMember(ourInfo: MemberInfo?, counterpartyInfo: MemberInfo): Boolean {
+        if (counterpartyInfo.isMgm) {
```
Comment (yift-r3) [inline components/link-manager/src/main/kotlin/net/corda/p2p/linkmanager/sessions/SessionManagerImpl.kt]: Maybe just `((counterpartyInfo.isMgm) || (ourInfo?.isMgm == true))`?

File: components/link-manager/src/main/kotlin/net/corda/p2p/linkmanager/sessions/SessionManagerImpl.kt [hunk 4]
```diff
@@ -286,6 +312,13 @@ internal class SessionManagerImpl(
         }
     }
 
+    private fun SessionCounterparties.calculateSessionMultiplicity(): Int {
+        if (communicationWithMgm) {
```
Comment (yift-r3) [inline components/link-manager/src/main/kotlin/net/corda/p2p/linkmanager/sessions/SessionManagerImpl.kt]: You can put the return before the if and add an else branch.

File: components/link-manager/src/nonOsgiIntegrationTest/kotlin/net/corda/p2p/P2PLayerEndToEndTest.kt [hunk 5]
```diff
@@ -516,7 +517,8 @@ class P2PLayerEndToEndTest {
                 .withValue(MAX_REPLAYING_MESSAGES_PER_PEER, ConfigValueFactory.fromAnyRef(100))
                 .withValue(HEARTBEAT_MESSAGE_PERIOD_KEY, ConfigValueFactory.fromAnyRef(Duration.ofSeconds(2)))
                 .withValue(SESSION_TIMEOUT_KEY, ConfigValueFactory.fromAnyRef(Duration.ofSeconds(10)))
-                .withValue(SESSIONS_PER_PEER_KEY, ConfigValueFactory.fromAnyRef(4))
+                .withValue(SESSIONS_PER_PEER_FOR_MEMBER_KEY, ConfigValueFactory.fromAnyRef(4))
```
Comment (yift-r3) [inline components/link-manager/src/nonOsgiIntegrationTest/kotlin/net/corda/p2p/P2PLayerEndToEndTest.kt]: 2?

File: components/link-manager/src/main/kotlin/net/corda/p2p/linkmanager/sessions/SessionManagerImpl.kt [hunk 6]
```diff
@@ -196,11 +203,18 @@ internal class SessionManagerImpl(
     @VisibleForTesting
     internal data class SessionManagerConfig(
         val maxMessageSize: Int,
-        val sessionsPerCounterparties: Int,
+        val sessionsPerCounterparties: Int?,
+        val numberOfSessionsPerPeer: NumberOfSessionsPerPeer,
```
Comment (yift-r3) [inline components/link-manager/src/main/kotlin/net/corda/p2p/linkmanager/sessions/SessionManagerImpl.kt]: Maybe we should have only the data class and build it from with either the `SESSIONS_PER_PEER_KEY` or the `SESSIONS_PER_PEER_FOR_MEMBER_KEY`, `SESSIONS_PER_PEER_FOR_MGM_KEY`?

File: components/link-manager/src/main/kotlin/net/corda/p2p/linkmanager/sessions/SessionManagerImpl.kt [hunk 7]
```diff
@@ -286,6 +329,16 @@ internal class SessionManagerImpl(
         }
     }
 
+    private fun SessionCounterparties.calculateSessionMultiplicity(): Int {
+        return if (config.get().sessionsPerCounterparties != null) {
+            config.get().sessionsPerCounterparties!!
```
Comment (yift-r3) [inline components/link-manager/src/main/kotlin/net/corda/p2p/linkmanager/sessions/SessionManagerImpl.kt]: You can avoid the !! by doing something like:
```
return config.get().sessionsPerCounterparties ?: if (communicationWithMgm) {
            config.get().numberOfSessionsPerPeer.forMgm
        } else {
            config.get().numberOfSessionsPerPeer.forMembers
        }
```

**Review by yift-r3 — COMMENTED** [review]
It looks OK to me. I left a couple of minor comments. You can choose to ignore them if you want.

**Review by yift-r3 — DISMISSED** [review]

**Review by yift-r3 — DISMISSED** [review]

**Review by dimosr — APPROVED** [review]

**Pull request merged**